### DISCUSSION TITLE
How can I use the API with Python requests module?

### DISCUSSION THREAD

**Discussion by CluelessTechnologist** [jesec/flood#381] [Q&A] [answered]
I couldn't find any documentation on the API other than [this](https://flood-api.netlify.app/). 
But that's JS right? I don't know JS...

How can I accomplish something like this:

```
requests.get(
  'http://flood.example.com:3000/api/torrents', 
  auth=HTTPBasicAuth('username', 'password')
)
```

I don't think basic auth is supported so any pointers on how to auth would be a good start 😄

**Comment by jesec** [comment]
APIs use JSON. 

You have to POST valid JSON to the appropriate endpoint. 

```
curl -X POST -H "Content-Type: application/json" -d '{"username":"test","password":"test"}' http://127.0.0.1:3000/api/auth/authenticate
```

Here is a simple curl command that does just that. It contains all info you need (at least for the authentication).

**Reply by CluelessTechnologist** [reply]
That works, but how can I now GET stuff from the API?

**Successful auth:**
```
curl -X POST -H "Content-Type: application/json" -d '{"username":"apiuser","password":"test"}' http://127.0.0.1:3000/api/auth/authenticate
{"success":true,"username":"apiuser","level":5`
```

**Trying to get a list of all torrents:**

```
curl -X GET -H 'Accept:application/json' http://127.0.0.1:3000/api/torrents
Unauthorized
```

**Reply by jesec** [reply]
You would need the cookie that contains authentication token. 

curl provides arguments that allow you to save and load cookies.

**Comment by CluelessTechnologist** [comment, marked as answer]
Reading up on how curl & cookies work I was able to build something that works in Python. Posting here in case someone else is looking for it:

```
#!/usr/bin/python3
import requests
curSession = requests.Session()

# all cookies received will be stored in the session object
payload={'username': "apiuser",'password': "test"}

resp1 = requests.post("http://127.0.0.1:3000/api/auth/authenticate", data=payload)

# manually pass previously returned cookies into following request
resp2 = requests.get("http://127.0.0.1:3000/api/torrents", cookies=resp1.cookies)
print(resp2.json())
```

**Reply by CluelessTechnologist** [reply]
I'm having some trouble with the API call for adding a torrent by URL though. Maybe you can help me out @jesec ?

```
curl -b cookie.txt -X POST -H "Content-Type: application/json" -d '{"urls": "https://releases.ubuntu.com/21.04/ubuntu-21.04-desktop-amd64.iso.torrent",  "cookies": "null", "destination": "/Downloads", "tags": "null", "isBasePath": "false", "isCompleted": "false", "isSequential": "false", "isInitialSeeding": "false", "start": "false"}' http://127.0.0.1:3000/api/torrents/add-urls
```

The response: 
```
{"message":"Validation error."}
```

I cannot figure out what I'm missing.
